pkg/rpc/client/wsclient.go

```diff
@@ -0,0 +1,145 @@
+package client
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/nspcc-dev/neo-go/pkg/rpc/request"
+	"github.com/nspcc-dev/neo-go/pkg/rpc/response"
+)
+
+// WSClient is a websocket-enabled RPC client that can be used with appropriate
+// servers. It's supposed to be faster than Client because it has persistent
+// connection to the server and at the same time is exposes some functionality
+// that is only provided via websockets (like event subscription mechanism).
+type WSClient struct {
+	Client
+	ws            *websocket.Conn
+	done          chan struct{}
+	notifications chan *request.In
+	responses     chan *response.Raw
+	requests      chan *request.Raw
+}
+
+// requestResponse is a combined type for request and response since we can get
+// any of them here.
+type requestResponse struct {
+	request.In
+	Error  *response.Error `json:"error,omitempty"`
+	Result json.RawMessage `json:"result,omitempty"`
+}
+
+const (
+	// Message limit for receiving side.
+	wsReadLimit = 10 * 1024 * 1024
+
+	// Disconnection timeout.
+	wsPongLimit = 60 * time.Second
+
+	// Ping period for connection liveness check.
+	wsPingPeriod = wsPongLimit / 2
+
+	// Write deadline.
+	wsWriteLimit = wsPingPeriod / 2
+)
+
+// NewWS returns a new WSClient ready to use (with established websocket
+// connection). You need to use websocket URL for it like `ws://1.2.3.4/ws`.
+func NewWS(ctx context.Context, endpoint string, opts Options) (*WSClient, error) {
+	cl, err := New(ctx, endpoint, opts)
+	cl.cli = nil
+
+	dialer := websocket.Dialer{HandshakeTimeout: opts.DialTimeout}
+	ws, _, err := dialer.Dial(endpoint, nil)
+	if err != nil {
+		return nil, err
+	}
+	wsc := &WSClient{
+		Client:    *cl,
+		ws:        ws,
+		responses: make(chan *response.Raw),
+		requests:  make(chan *request.Raw),
+		done:      make(chan struct{}),
+	}
+	go wsc.wsReader()
```

Comment:
Can `wsc` be stopped manually?

Reply:
Yeah, I've forgot to add `Close()` in this patchset, probably it's worth doing so right here rather than postpone that for another set.